File: earth_moon_sim.py
from vpython import sphere, vector, rate, color, scene, textures
import numpy as np

#constants
G   = 6.67430e-11          # gravitational constant
M_e = 5.972e24             # Earth mass (kg)
M_m = 7.3477e22            # Moon mass  (kg)
r_em= 384400e3             # Earth–Moon distance (m)
T_m = 27.322 * 24*3600     # Moon orbital period (s)
ω   = 2*np.pi / T_m        # Moon angular speed

#scene setup
init_pos = vector(6671e3, 0, 0)   # start 200 km above Earth
init_vel = vector(0, 7800, 0)     # approx LEO speed
dt       = 60                     # simulation timestep (s)
t_max    = 5 * 24*3600            # run for 5 days

scene.width  = 800
scene.height = 600
scene.title  = "Earth–Moon–Spacecraft Simulation"


#earth eith texture
earth = sphere(
    pos=vector(0,0,0),
    radius=6.371e6,
    texture=textures.earth,
    shininess=0.8
)

# Starfield background
star_layer = box(
    pos=vector(0, 0, -1e9),
    size=vector(3e9, 3e9, 1e6),
    texture="https://upload.wikimedia.org/wikipedia/commons/0/01/Starsinthesky.jpg",
    emissive=True,
    opacity=1
)

moon = sphere(
    pos=vector(r_em,0,0),
    radius=1.737e6,
    color=color.white,
    make_trail=True,
    trail_radius=1e6
)

#spacecreaft with trail 
craft = sphere(
    pos=init_pos,
    radius=2e6,
    color=color.red,
    make_trail=True,
    trail_radius=5e5
)


def moon_pos(t):
    return vector(r_em*np.cos(ω*t),
                  r_em*np.sin(ω*t),
                  0)

def acceleration(r, t):
    # earthgrav
    a_e = G * M_e * (-r) / (r.mag**3)
    # moongrav
    r_m = moon_pos(t) - r
    a_m = G * M_m * r_m / (r_m.mag**3)
    return a_e + a_m

# AI below - inital state 
r = init_pos
v = init_vel
t = 0.0

#simulation loop 
while t < t_max:
    rate(100)   # cap to 100 iterations/sec

    # move Moon
    moon.pos = moon_pos(t)

    # RK4 integration for the craft
    k1v = acceleration(r, t) * dt
    k1r = v * dt

    k2v = acceleration(r + 0.5*k1r, t + 0.5*dt) * dt
    k2r = (v + 0.5*k1v) * dt

    k3v = acceleration(r + 0.5*k2r, t + 0.5*dt) * dt
    k3r = (v + 0.5*k2v) * dt

    k4v = acceleration(r + k3r, t + dt) * dt
    k4r = (v + k3v) * dt

    v = v + (k1v + 2*k2v + 2*k3v + k4v) / 6
    r = r + (k1r + 2*k2r + 2*k3r + k4r) / 6

    craft.pos = r
    t += dt

File: earth_moon_sim.cpp
#include <SFML/Graphics.hpp>
#include <cmath>
#include <iostream>

const double G   = 6.67430e-11;
const double M_s = 1.989e30;         // Sun mass (kg)
const double M_e = 5.972e24;         // Earth mass (kg)
const double M_m = 7.3477e22;        // Moon mass (kg)
const double r_se= 1.496e11;         // Sun–Earth distance (m)
const double r_em= 384400e3;         // Earth–Moon distance (m)
const double T_e = 365.25 * 24*3600; // Earth orbital period (s)
const double T_m = 27.322 * 24*3600; // Moon orbital period (s)
const double omega_e = 2*M_PI / T_e;
const double omega_m = 2*M_PI / T_m;

const double dt    = 60;              // simulation timestep (s)
const double t_max = 5 * 24*3600;     // run for 5 days

// Visualization scale (meters to pixels)
const double SCALE = 2.5e-9;

struct Vec2 {
    double x, y;
    Vec2 operator+(const Vec2& v) const { return {x+v.x, y+v.y}; }
    Vec2 operator-(const Vec2& v) const { return {x-v.x, y-v.y}; }
    Vec2 operator*(double s) const { return {x*s, y*s}; }
    Vec2 operator/(double s) const { return {x/s, y/s}; }
    double mag() const { return std::sqrt(x*x + y*y); }
};

Vec2 sun_pos() {
    return {0, 0};
}

Vec2 earth_pos(double t) {
    return {r_se * std::cos(omega_e * t), r_se * std::sin(omega_e * t)};
}

Vec2 moon_pos(double t) {
    Vec2 e = earth_pos(t);
    return e + Vec2{r_em * std::cos(omega_m * t), r_em * std::sin(omega_m * t)};
}

Vec2 acceleration(const Vec2& r, double t) {
    // Sun gravity
    Vec2 a_s = (sun_pos() - r) * (G * M_s / std::pow((sun_pos() - r).mag(), 3));
    // Earth gravity
    Vec2 e = earth_pos(t);
    Vec2 a_e = (e - r) * (G * M_e / std::pow((e - r).mag(), 3));
    // Moon gravity
    Vec2 m = moon_pos(t);
    Vec2 a_m = (m - r) * (G * M_m / std::pow((m - r).mag(), 3));
    return a_s + a_e + a_m;
}

int main() {
    sf::RenderWindow window(sf::VideoMode(1200, 900), "Sun-Earth-Moon-Spacecraft Simulation");
    window.setFramerateLimit(60);

    // Initial spacecraft state (200 km above Earth, in Earth's direction)
    double t = 0.0;
    Vec2 r = earth_pos(0) + Vec2{6671e3, 0};
    Vec2 v = {0, 7800};

    // For trails
    sf::VertexArray craftTrail(sf::LineStrip);

    while (window.isOpen() && t < t_max) {
        sf::Event event;
        while (window.pollEvent(event))
            if (event.type == sf::Event::Closed) window.close();

        // RK4 integration
        Vec2 k1v = acceleration(r, t) * dt;
        Vec2 k1r = v * dt;

        Vec2 k2v = acceleration(r + k1r * 0.5, t + 0.5*dt) * dt;
        Vec2 k2r = (v + k1v * 0.5) * dt;

        Vec2 k3v = acceleration(r + k2r * 0.5, t + 0.5*dt) * dt;
        Vec2 k3r = (v + k2v * 0.5) * dt;

        Vec2 k4v = acceleration(r + k3r, t + dt) * dt;
        Vec2 k4r = (v + k3v) * dt;

        v = v + (k1v + k2v*2 + k3v*2 + k4v) / 6.0;
        r = r + (k1r + k2r*2 + k3r*2 + k4r) / 6.0;
        t += dt;

        // Draw
        window.clear(sf::Color::Black);

        // Sun
        sf::CircleShape sun(20);
        sun.setFillColor(sf::Color::Yellow);
        sun.setOrigin(20, 20);
        sun.setPosition(600, 450);
        window.draw(sun);

        // Earth
        Vec2 e = earth_pos(t);
        sf::CircleShape earth(8);
        earth.setFillColor(sf::Color::Blue);
        earth.setOrigin(8, 8);
        earth.setPosition(600 + e.x * SCALE, 450 + e.y * SCALE);
        window.draw(earth);

        // Moon
        Vec2 m = moon_pos(t);
        sf::CircleShape moon(4);
        moon.setFillColor(sf::Color(200, 200, 200));
        moon.setOrigin(4, 4);
        moon.setPosition(600 + m.x * SCALE, 450 + m.y * SCALE);
        window.draw(moon);

        // Spacecraft
        sf::CircleShape craft(3);
        craft.setFillColor(sf::Color::Red);
        craft.setOrigin(3, 3);
        craft.setPosition(600 + r.x * SCALE, 450 + r.y * SCALE);
        window.draw(craft);

        // Trail
        craftTrail.append(sf::Vertex(sf::Vector2f(600 + r.x * SCALE, 450 + r.y * SCALE), sf::Color::Green));
        window.draw(craftTrail);

        window.display();
    }
    return 0;
}